File: android-sdui/app/src/main/java/com/hsbc/sdui/engine/SDUIStaticDistribution.kt
// SDUIStaticDistribution.kt
// HSBC SDUI Android Renderer
//
// Implements the three-tier resolution chain for SDUI screens:
//   1. Fetch manifest.json from CDN — compare version against SharedPreferences
//   2a. Version unchanged → serve local file cache (no download)
//   2b. Version changed / no local file → download, SHA-256 verify, persist, render
//   3. CDN unreachable → serve local file cache; if absent → serve bundled baseline
//
// Also manages self-pick entry-point preferences:
//   • Customer preferences stored in SharedPreferences on-device
//   • Preferences preserved across remote updates unless selfPickForceUpdate == true
//
// All I/O is performed on IO dispatcher — never block the main thread.

package com.hsbc.sdui.engine

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.hsbc.sdui.analytics.AnalyticsClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.net.URL
import java.security.MessageDigest

// ─── Manifest types ──────────────────────────────────────────────────────────

data class SDUIManifest(
    val schemaVersion: String,
    val generatedAt: String,
    val screens: Map<String, SDUIScreenEntry>,
    val selfPickForceUpdate: Boolean = false
)

data class SDUIScreenEntry(
    val version: String,
    val etag: String,
    val sizeBytes: Int
)

data class SelfPickEntry(
    val items: List<Map<String, String>>,
    val savedAt: String
)

// ─── SDUIStaticDistribution ───────────────────────────────────────────────────

class SDUIStaticDistribution(
    private val context: Context,
    private val cdnBase: String  = "https://cdn.hsbc.com/sdui",
    private val appId: String    = "hsbcmobile",
    private val platform: String = "android",
    private val userId: String,
) {
    private val gson = Gson()

    // File cache directory: <cacheDir>/SDUI/
    private val cacheDir: File = File(context.cacheDir, "SDUI").also { it.mkdirs() }

    // SharedPreferences store (replaces DataStore — no extra dependency needed)
    private val prefs: SharedPreferences =
        context.getSharedPreferences("sdui_static", Context.MODE_PRIVATE)

    private fun versionKey(screenId: String) = "sdui_version_$screenId"
    private fun selfPickKey(screenId: String) = "selfPick_${userId}_$screenId"

    // ── Public API ────────────────────────────────────────────────────────────

    /**
     * Resolves the best available SDUI JSON string for [screenId].
     * Returns Pair(json, isStale) — isStale=true means we served local/bundled.
     */
    suspend fun resolve(screenId: String): Pair<String?, Boolean> = withContext(Dispatchers.IO) {
        // Step 1 — try CDN manifest
        val manifest = fetchManifest()
        if (manifest != null) {
            val entry = manifest.screens[screenId]
            if (entry != null) {
                val storedVersion = prefs.getString(versionKey(screenId), null)
                if (storedVersion == entry.version) {
                    val local = loadLocalFile(screenId)
                    if (local != null) return@withContext Pair(local, false)
                }
                // Version changed or no local file — download from CDN
                val downloaded = downloadScreen(screenId, entry.version, entry.etag)
                if (downloaded != null) {
                    persistLocalFile(screenId, downloaded, entry.version)
                    return@withContext Pair(downloaded, false)
                }
                // Download failed (corrupted / missing on CDN)
                AnalyticsClient.fire(
                    "sdui_cdn_download_failed",
                    mapOf("screen_id" to screenId, "version" to entry.version)
                )
            }
        }

        // Step 3 — CDN unreachable or no entry
        val local = loadLocalFile(screenId)
        if (local != null) return@withContext Pair(local, true)

        // Last resort — bundled baseline asset
        val bundled = loadBundledBaseline(screenId)
        Pair(bundled, true)
    }

    /**
     * Returns the self-pick items for the SELF_PICK_ENTRY_POINTS slice.
     * Clears saved prefs and returns remote defaults when [forceUpdate] is true.
     */
    suspend fun resolvedSelfPickItems(
        screenId: String,
        remoteDefaults: List<Map<String, String>>,
        forceUpdate: Boolean
    ): List<Map<String, String>> = withContext(Dispatchers.IO) {
        val key = selfPickKey(screenId)
        if (forceUpdate) {
            prefs.edit().remove(key).apply()
            return@withContext remoteDefaults
        }
        val raw = prefs.getString(key, null) ?: return@withContext remoteDefaults
        val type = object : TypeToken<SelfPickEntry>() {}.type
        val entry: SelfPickEntry? = runCatching {
            gson.fromJson<SelfPickEntry>(raw, type)
        }.getOrNull()
        entry?.items ?: remoteDefaults
    }

    /** Persists customer self-pick ordering to SharedPreferences. */
    suspend fun saveCustomerSelfPick(screenId: String, items: List<Map<String, String>>) {
        withContext(Dispatchers.IO) {
            val entry = SelfPickEntry(items = items, savedAt = System.currentTimeMillis().toString())
            prefs.edit().putString(selfPickKey(screenId), gson.toJson(entry)).apply()
        }
    }

    // ── Manifest ─────────────────────────────────────────────────────────────

    private fun fetchManifest(): SDUIManifest? = runCatching {
        val text = URL("$cdnBase/$appId/$platform/manifest.json").readText(Charsets.UTF_8)
        gson.fromJson(text, SDUIManifest::class.java)
    }.getOrNull()

    // ── CDN Download ──────────────────────────────────────────────────────────

    private fun downloadScreen(screenId: String, version: String, etag: String): String? =
        runCatching {
            val bytes = URL("$cdnBase/$appId/$platform/$screenId/$version.json").readBytes()

            // SHA-256 verify against etag (stored as hex-encoded hash, possibly quoted)
            val stripped = etag.replace("\"", "")
            val digest   = MessageDigest.getInstance("SHA-256").digest(bytes)
            val computed = digest.joinToString("") { "%02x".format(it) }
            if (computed.lowercase() != stripped.lowercase()) {
                AnalyticsClient.fire(
                    "sdui_integrity_fail",
                    mapOf("screen_id" to screenId, "expected" to stripped, "got" to computed)
                )
                return null
            }
            String(bytes, Charsets.UTF_8)
        }.getOrNull()

    // ── Local File Cache ──────────────────────────────────────────────────────

    private fun localFile(screenId: String) = File(cacheDir, "$screenId.json")

    private fun loadLocalFile(screenId: String): String? =
        runCatching { localFile(screenId).readText(Charsets.UTF_8) }.getOrNull()

    private fun persistLocalFile(screenId: String, json: String, version: String) {
        runCatching { localFile(screenId).writeText(json, Charsets.UTF_8) }
        prefs.edit().putString(versionKey(screenId), version).apply()
    }

    // ── Bundled Baseline ──────────────────────────────────────────────────────

    private fun loadBundledBaseline(screenId: String): String? = runCatching {
        // Baselines shipped as raw resource files: res/raw/sdui_baseline_{screenId}.json
        val resName = "sdui_baseline_${screenId.replace('-', '_')}"
        val resId   = context.resources.getIdentifier(resName, "raw", context.packageName)
        if (resId == 0) return null
        context.resources.openRawResource(resId).bufferedReader().readText()
    }.getOrNull()
}

File: ios-sdui/Sources/HSBC_SDUI/Engine/SDUIStaticDistribution.swift
// SDUIStaticDistribution.swift
// HSBC SDUI iOS Renderer
//
// Implements the three-tier resolution chain for SDUI screens:
//   1. Fetch manifest.json from CDN — compare version against local storage
//   2a. Version unchanged → serve local storage (no download)
//   2b. Version changed / no local file → download, SHA-256 verify, persist, render
//   3. CDN unreachable → serve local storage; if absent → serve bundled baseline
//
// Also manages self-pick entry-point customisation:
//   • Customer preferences stored in UserDefaults under "selfPick_{userId}_{screenId}"
//   • Preferences are preserved across remote updates unless selfPickForceUpdate == true

import Foundation
import CryptoKit

// MARK: - Manifest types

struct SDUIManifest: Codable {
    let schemaVersion: String
    let generatedAt: String
    let screens: [String: SDUIScreenEntry]
    let selfPickForceUpdate: Bool

    enum CodingKeys: String, CodingKey {
        case schemaVersion      = "schemaVersion"
        case generatedAt        = "generatedAt"
        case screens
        case selfPickForceUpdate = "selfPickForceUpdate"
    }
}

struct SDUIScreenEntry: Codable {
    let version: String
    let etag: String
    let sizeBytes: Int
}

// MARK: - SelfPickEntry

struct SelfPickEntry: Codable {
    let items: [[String: String]]   // ordered list of entry-point ids + labels
    let savedAt: String             // ISO-8601
}

// MARK: - SDUIStaticDistribution

/// Encapsulates the full CDN manifest-check → local-storage → bundled-baseline resolution
/// chain and the self-pick preference layer. Thread-safe; all work done on a background
/// actor — callers switch back to MainActor before updating UI.
actor SDUIStaticDistribution {

    // MARK: Configuration

    private let cdnBase: String
    private let appId: String
    private let platform: String
    private let userId: String
    private let session: URLSession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    // File-cache directory: <Caches>/SDUI/
    private let cacheDir: URL

    // UserDefaults key prefixes
    private let localVersionPrefix = "hsbc_sdui_version_"
    private let selfPickPrefix     = "selfPick_"

    // MARK: Init

    init(
        cdnBase: String  = "https://cdn.hsbc.com/sdui",
        appId: String    = "hsbcmobile",
        platform: String = "ios",
        userId: String,
        session: URLSession = .shared
    ) {
        self.cdnBase   = cdnBase
        self.appId     = appId
        self.platform  = platform
        self.userId    = userId
        self.session   = session

        let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        self.cacheDir = base.appendingPathComponent("SDUI", isDirectory: true)
        try? FileManager.default.createDirectory(at: cacheDir, withIntermediateDirectories: true)
    }

    // MARK: - Public API

    /// Resolves the best available SDUI JSON data for `screenId` using the 3-tier chain.
    /// Returns `(data, isStale)` where `isStale` is true when falling back to local/bundled.
    func resolve(screenId: String) async -> (Data?, isStale: Bool) {
        // Step 1 — try CDN manifest
        if let manifest = await fetchManifest() {
            if let entry = manifest.screens[screenId] {
                let localVersion = storedVersion(for: screenId)
                if localVersion == entry.version, let local = loadLocalFile(screenId: screenId) {
                    // Version unchanged — serve local storage (no download)
                    return (local, false)
                }
                // Version changed or no local file — download from CDN
                if let downloaded = await downloadScreen(screenId: screenId, version: entry.version, etag: entry.etag) {
                    persistLocalFile(screenId: screenId, data: downloaded, version: entry.version)
                    return (downloaded, false)
                }
                // Download failed (corrupted / missing) — keep existing local
                AnalyticsClient.fire(
                    event: "sdui_cdn_download_failed",
                    properties: ["screen_id": screenId, "version": entry.version]
                )
            }
        }

        // Step 3 — CDN unreachable or no screen entry
        if let local = loadLocalFile(screenId: screenId) {
            return (local, true)
        }

        // Last resort — bundled baseline
        if let bundled = loadBundledBaseline(screenId: screenId) {
            return (bundled, true)
        }

        return (nil, true)
    }

    // MARK: - Self-Pick

    /// Returns merged self-pick items for the `SELF_PICK_ENTRY_POINTS` slice.
    /// Remote defaults are used as-is when `forceUpdate` is true (and saved prefs are cleared).
    /// Otherwise, saved customer preferences override the remote defaults.
    func resolvedSelfPickItems(
        screenId: String,
        remoteDefaults: [[String: String]],
        forceUpdate: Bool
    ) -> [[String: String]] {
        let key = selfPickKey(screenId: screenId)

        if forceUpdate {
            // Clear saved preferences — customer sees new defaults, can re-customise
            UserDefaults.standard.removeObject(forKey: key)
            return remoteDefaults
        }

        guard
            let data = UserDefaults.standard.data(forKey: key),
            let saved = try? decoder.decode(SelfPickEntry.self, from: data)
        else {
            return remoteDefaults
        }

        return saved.items
    }

    /// Persists customer self-pick preferences to UserDefaults.
    func saveCustomerSelfPick(screenId: String, items: [[String: String]]) {
        let entry = SelfPickEntry(items: items, savedAt: ISO8601DateFormatter().string(from: Date()))
        guard let data = try? encoder.encode(entry) else { return }
        UserDefaults.standard.set(data, forKey: selfPickKey(screenId: screenId))
    }

    // MARK: - Manifest

    private func fetchManifest() async -> SDUIManifest? {
        guard let url = URL(string: "\(cdnBase)/\(appId)/\(platform)/manifest.json") else { return nil }
        var request = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10)
        request.httpMethod = "GET"

        do {
            let (data, response) = try await session.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }
            return try decoder.decode(SDUIManifest.self, from: data)
        } catch {
            return nil
        }
    }

    // MARK: - CDN Screen Download

    private func downloadScreen(screenId: String, version: String, etag: String) async -> Data? {
        guard let url = URL(string: "\(cdnBase)/\(appId)/\(platform)/\(screenId)/\(version).json") else {
            return nil
        }

        do {
            let (data, response) = try await session.data(for: URLRequest(url: url))
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }

            // SHA-256 verify against etag (etag stored as hex-encoded SHA-256)
            let stripped = etag.trimmingCharacters(in: CharacterSet(charactersIn: "\""))
            let digest   = SHA256.hash(data: data)
            let computed = digest.compactMap { String(format: "%02x", $0) }.joined()
            guard computed.lowercased() == stripped.lowercased() else {
                AnalyticsClient.fire(
                    event: "sdui_integrity_fail",
                    properties: ["screen_id": screenId, "expected": stripped, "got": computed]
                )
                return nil
            }
            return data
        } catch {
            return nil
        }
    }

    // MARK: - Local File Cache

    private func localFileURL(screenId: String) -> URL {
        cacheDir.appendingPathComponent("\(screenId).json")
    }

    private func loadLocalFile(screenId: String) -> Data? {
        try? Data(contentsOf: localFileURL(screenId: screenId))
    }

    private func persistLocalFile(screenId: String, data: Data, version: String) {
        try? data.write(to: localFileURL(screenId: screenId), options: .atomic)
        UserDefaults.standard.set(version, forKey: localVersionPrefix + screenId)
    }

    private func storedVersion(for screenId: String) -> String? {
        UserDefaults.standard.string(forKey: localVersionPrefix + screenId)
    }

    // MARK: - Bundled Baseline

    private func loadBundledBaseline(screenId: String) -> Data? {
        // Baselines shipped as JSON files in the app bundle under Resources/SDUIBaseline/
        guard let url = Bundle.main.url(forResource: screenId, withExtension: "json", subdirectory: "SDUIBaseline") else {
            return nil
        }
        return try? Data(contentsOf: url)
    }

    // MARK: - Helpers

    private func selfPickKey(screenId: String) -> String {
        "\(selfPickPrefix)\(userId)_\(screenId)"
    }
}
